### ISSUE TITLE
Disconnection detection issue

### ISSUE THREAD

**Issue by ScreamZ** [Links2004/arduinoWebSockets#864]
Hi,

I'm working on a disconnect scenario to improve the documentation and my workflow.

My computer is running a WebSocket server.
The ESP32 is the client. (`WebSocketsClient webSocket;`)

## 1 - Disconnected from server shutdown

✅ Seems to work properly and properly detected by the library.

## 2 - Switching the network from the server

To simulate disconnection, I also switched off the WiFi on my computer, so the server is no longer reachable.
**This is an edge case I guess** but why can't this happen?

❌ **The ESP32 with socket doesn't detect the disconnection.** Within `onEvent` -> `WStype_DISCONNECTED` 

### Current workaround

Currently, I'm pinging frequently the server and if the last message is > 10 seconds, I consider it off.

Do you think the library can handle this scenario of failure without having to rely on custom code for pings?

**Comment by Links2004** [comment]
the ESP tcp socket only notices that the socket is dead when something is send.
if you try to send something or enableHeartbeat sends its ping. 

https://github.com/Links2004/arduinoWebSockets/blob/6a26f74f958b7d31d71442d59e376a98cb53bedd/src/WebSocketsClient.h#L102-L102

after a send failes the client disconnects and the event for  `WStype_DISCONNECTED` is send.

backround:
if you hard disconnect the network on the servers side no FIN/ACK from the server side can be send and as a result the ESP TCP stack never gets the "info" that the server is gone.
we are left only with the ACK timout on TX to detect this if the server is simply gone and does not answer any more.

**Comment by AviStudio** [comment]
Tell me please, how long can a WebSocket Client connection be? I would like to send some sensor data over it in JSON, and the scenario is that sensor should be online 24/7

**Comment by Links2004** [comment]
there is no limit on the arduinoWebSockets side, it all depends on how your server is setup and how stable the WiFi / Network connection is.
I my self use this lib in home automation tasks and with the heartbeat enabled I had never problems with long dead connections.

**Comment by AviStudio** [comment]
I would be very happy if you could share an example how to use heartbeat.
In another way I would like to ask if there are some possible limits or problems if I'll send data over WebSocket in JSON format, maybe I should do some RAM cleanup or something else?

**Comment by ScreamZ** [comment]
> I would be very happy if you could share an example how to use heartbeat. In another way I would like to ask if there are some possible limits or problems if I'll send data over WebSocket in JSON format, maybe I should do some RAM cleanup or something else?

I could only recommend using https://arduinojson.org/, on ESP32 there is no real RAM concern as they have a lot.

**Comment by Links2004** [comment]
yes using JSON is works with ESP8266 and ESP32 fine.

e.g. 
```
            DynamicJsonDocument doc(1024);
            DeserializationError error = deserializeJson(doc, payload, length);
            if(error) {
                USE_SERIAL.print(F("deserializeJson() failed: "));
                USE_SERIAL.println(error.c_str());
                return;
            }

```

heartbeat only needs to be enabled.

https://github.com/Links2004/arduinoWebSockets/blob/93707d455fff3c667aedd92b485276101f5e6eba/src/WebSocketsClient.h#L102

**Comment by AviStudio** [comment]
Hello again. Here i have a few questions:
1. Where should i call client.enableHeartbeat()?  in "setup" will be ok? on in "loop" ? 
2. pingInterval, pongTimeout are in milliseconds?

**Comment by Links2004** [comment]
yes client.enableHeartbeat is called in the setup process, and yes all times are in milliseconds.

https://github.com/Links2004/arduinoWebSockets/blob/93707d455fff3c667aedd92b485276101f5e6eba/src/WebSocketsClient.cpp#L962-L970

**Comment by Links2004** [comment]
any reason to not use `enableHeartbeat` there too?

https://github.com/Links2004/arduinoWebSockets/blob/93707d455fff3c667aedd92b485276101f5e6eba/src/WebSocketsServer.h#L90-L90

each side can send a ping, the ws protocol allows it. 

https://datatracker.ietf.org/doc/html/rfc6455#section-5.5.2

**Comment by slowpoison4** [comment]
Hi Links2004,
ive been using the enable heartBeat for ESP32 for my ws client. What i notice is that randomly the websocket crashes (which is fine) but then i notice the Heap memory starts to decline quite rapidly. When i removed the enable heartbeat and manually shut of the internet using a mobile hotspot, i didnt notice any heap memory reduction. It only happens when using enable heartbeat. Also further more, when the reconnection does happen, the lost heap memory isnt returned and eventually the heap memory becomes so low that i get no more process for wifi and then it crashes. My current workaround is to restart the esp when a disconnection is detected. This is less than ideal as this could cause missing some sensor data during the intial setup phase. Any suggestions as to why this is happening?

**Comment by prodjallen** [comment]
Same Issue, with or without heartbeat, disconnection isn't detected on server side, even on tight intervals. Sometimes it is detected as a disconnect, after 300 sec. but not always. I am sending a small amount of data every 30-50ms, however if another esp32 is connected as a client, abruptly disconnects ie power disconnect etc, then its not detected and the socket server becomes frozen, nothing outputs on websockets, however the async webserver still serves clients the webpage without issue, only the websocket freezes up.

**Comment by AviStudio** [comment]
Hello again! Using this library for more than 6 months w/o problems, now I'm developing another project, ESP32 WS Client and NodeJS WS Server, want to ask, are there some recommended arguments values for enableHeartbeat()?
Tried to increase: webSocket.enableHeartbeat(10000, 5000, 3), but got multiple `WStype_DISCONNECTED` events :(
Now i'm running webSocket.enableHeartbeat(2000, 1000, 3); Is this OK?

**Comment by ScreamZ** [comment]
> Hello again! Using this library for more than 6 months w/o problems, now I'm developing another project, ESP32 WS Client and NodeJS WS Server, want to ask, are there some recommended arguments values for `enableHeartbeat()`? Tried to increase:` webSocket.enableHeartbeat(10000, 5000, 3)`, but got multiple `WStype_DISCONNECTED` events :( Now i'm running `webSocket.enableHeartbeat(2000, 1000, 3)`; Is this a good practice?
> 
> P.S: I'm sending data to WS Server every 6000ms, so maybe there is not necessary to use enableHeartbeat() ?

Heartbeat is a mechanism that allows to maintain the connection alive, you need to send data on a regular basis depending on the server requirements, but sometime the server sent a ping and the client respond is enough, if you send data regularly you might not need heartbeat.
